{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2>Biopython (deel 2) en BioMart</h2>\n",
    "\n",
    "Gebruik voor opdracht 1 de Modules Bio.AlignIO en Bio.Align.AlignInfo. Documentatie en voorbeelden kun je vinden op: http://biopython.org/DIST/docs/tutorial/Tutorial.html.\n",
    "\n",
    "<b>Opdracht 1: Alignments</b>\n",
    "\n",
    "Het bestand 'clustalw.aln' bevat een alignment in clustal format.\n",
    "\n",
    "a) Zet het bestand om in fasta-format en sla dit op als 'fasta.aln'.<br>\n",
    "b) Print de lengte van de alignment.<br>\n",
    "c) Maak een Position-Specific Score Matrix van de alignment.<br>\n",
    "d) Gebruik de PSSM om te kijken op welke posities één van de basen 18x of vaker voorkomt. Print hiervan de positie, de base en het aantal keer dat de base op de positie voorkomt. De output moet er als volgt uit komen te zien:<br>\n",
    "```\n",
    "29 A 18\n",
    "30 T 22\n",
    "...\n",
    "67 A 18\n",
    "100 G 18\n",
    "```\n",
    "e) Maak een consensus sequentie van de alignment en print hiervan posities 28 t/m 48.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Lengte van de alignment: 117\n",
      "29 A 18\n",
      "30 T 22\n",
      "31 G 22\n",
      "32 T 22\n",
      "34 A 22\n",
      "37 A 19\n",
      "38 A 18\n",
      "42 T 22\n",
      "44 A 22\n",
      "45 C 22\n",
      "46 A 21\n",
      "47 T 22\n",
      "52 T 18\n",
      "57 A 19\n",
      "67 A 18\n",
      "100 G 18\n",
      "De consensus sequentie is:  XATGTXAXXAAXXCTXACATX\n"
     ]
    }
   ],
   "source": [
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "\n",
    "# a)\n",
    "AlignIO.convert(\"clustalw.aln\", \"clustal\", \"fasta.aln\", \"fasta\")\n",
    "\n",
    "# b) Voor onderstaande opdrachten kun je ook verder gaan met clustalw.aln.\n",
    "fasta_alignment = AlignIO.read(open(\"fasta.aln\"), \"fasta\")\n",
    "\n",
    "alignment_length = fasta_alignment.get_alignment_length()\n",
    "print(\"Lengte van de alignment:\", alignment_length)\n",
    "\n",
    "# c)\n",
    "summary_align = AlignInfo.SummaryInfo(fasta_alignment)\n",
    "\n",
    "# pos_specific_score_matrix lijkt outdated en geeft een waarschuwing. Die mag je voor nu negeren.\n",
    "from warnings import filterwarnings\n",
    "filterwarnings(\"ignore\")\n",
    "\n",
    "pssm = summary_align.pos_specific_score_matrix()\n",
    "# print(pssm)\n",
    "\n",
    "# d)\n",
    "for pos in range(alignment_length):\n",
    "    for base in [\"A\", \"C\", \"G\", \"T\"]:\n",
    "        if pssm[pos][base] >= 18:\n",
    "            print(pos+1, base, int(pssm[pos][base]))\n",
    "\n",
    "# e)\n",
    "consensus = summary_align.dumb_consensus()\n",
    "print(\"De consensus sequentie is: \", consensus[27:48]) #28 t/m 48\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b>Opdracht 2: BioMart</b>\n",
    "\n",
    "Dit script maakt gebruik van de package [_biomart_](https://pypi.python.org/pypi/biomart). Deze is __niet__ voor-geinstalleerd op anaconda. Je kunt *biomart* installeren met behulp van pip (een package management system om Python software packages te installeren en managen). Open de Anaconda Prompt en typ \"```pip install biomart```\".\n",
    "\n",
    "In deze opdracht gaan we [Ensembl](https://www.ensembl.org) gebruiken.\n",
    "\n",
    "Hieronder een voorbeeld hoe je van het gen met ENSG00000085840 de upstream region (het gedeelte voor het gen; de promoter) kan ophalen."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene stable ID\n",
      "ENSG00000085840\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from biomart import BiomartServer\n",
    "\n",
    "server = BiomartServer(\"http://www.ensembl.org/biomart\")\n",
    "human_dataset = server.datasets['hsapiens_gene_ensembl']\n",
    "\n",
    "\"\"\"\n",
    "response = human_dataset.search({\n",
    "  'filters' : {'ensembl_gene_id' : 'ENSG00000085840', 'upstream_flank' : 200},\n",
    "  'attributes' : ['ensembl_gene_id', 'gene_flank']\n",
    "  }, header=1)\n",
    "print(response.text)\n",
    "\"\"\"\n",
    "\n",
    "## Update:\n",
    "## Het kan zijn dat de upstream flank soms een foutmelding geeft.\n",
    "## In dat geval een paar keer de code runnen om te kijken of het een keer goed gaat.\n",
    "## Blijft het fout gaan dan bij de filters de upstream flank weghalen\n",
    "## en bij de attributes de gene flank weghalen.\n",
    "## Bij de opdracht hoeft dan de upstream flank niet getoond te worden.\n",
    "\n",
    "\n",
    "## Werkende code zonder upstream flank\n",
    "response = human_dataset.search({\n",
    "  'filters' : {'ensembl_gene_id' : 'ENSG00000085840'},\n",
    "  'attributes' : ['ensembl_gene_id']\n",
    "  }, header=1)\n",
    "print(response.text)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Een dataset beschikt over filters en attributen. Met filters kan je aangeven aan welke criteria je data moet voldoen. Veelgebruikte filters zijn bijvoorbeeld gennamen of eiwit identifiers. De volledige lijst met beschikbare filters kun je vinden door de dictionary ```<datasetnaam>.show_filters()``` te bekijken. De volledige lijst met attributen kun je bekijken met ```<datasetnaam>.show_attributes()```.\n",
    "\n",
    "* Gebruik de volgende filters:\n",
    "    1. Ensembl genen: ENSG00000123080, ENSG00000116574, ENSG00000125885, ENSG00000129355, ENSG00000183434, ENSG00000172201, ENSG00000206557, ENSG00000204086\n",
    "    2. Upstream flank van 200 bp.\n",
    "* En toon de volgende attributen:\n",
    "    1. Ensembl gene ID\n",
    "    2. Strand waarop het gen zit\n",
    "    3. Startpositie van het gen\n",
    "    4. Stoppositie van het gen\n",
    "    5. Transcript startpositie\n",
    "    6. 'Genomic coding' startpositie\n",
    "    7. Upstream flank van 200 bp \n",
    "\n",
    "De response van de query komt terug in de vorm van tab-delimited text."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene stable ID\tStrand\tGene start (bp)\tGene end (bp)\tTranscript start (bp)\tGenomic coding start\n",
      "ENSG00000116574\t1\t228735479\t228746664\t228735479\t228735743\n",
      "ENSG00000116574\t1\t228735479\t228746664\t228735479\t228737673\n",
      "ENSG00000116574\t1\t228735479\t228746664\t228735479\t228743285\n",
      "ENSG00000123080\t1\t50960745\t50974634\t50960745\t\n",
      "ENSG00000123080\t1\t50960745\t50974634\t50960745\t50970369\n",
      "ENSG00000123080\t1\t50960745\t50974634\t50960745\t50973893\n",
      "ENSG00000123080\t1\t50960745\t50974634\t50968706\t50970369\n",
      "ENSG00000123080\t1\t50960745\t50974634\t50968706\t\n",
      "ENSG00000123080\t1\t50960745\t50974634\t50968706\t50973893\n",
      "ENSG00000123080\t1\t50960745\t50974634\t50970245\t50973893\n",
      "ENSG00000123080\t1\t50960745\t50974634\t50970245\t50970369\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950652\t\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950652\t5952016\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950652\t5952424\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950652\t5954608\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950652\t5955102\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950652\t5957126\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950652\t5958528\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950652\t5963274\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950652\t5967436\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950652\t5967830\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950652\t5972007\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950652\t5977876\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950652\t5982970\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950652\t5984781\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950652\t5985922\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950652\t5987282\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950652\t5993506\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950652\t5994299\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950652\t5972576\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950652\t5973056\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950931\t5952016\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950931\t5952424\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950931\t5954608\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950931\t5955102\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950931\t5957126\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950931\t5958528\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950931\t5963274\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950931\t5967436\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950931\t5972007\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950931\t5977876\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950931\t5982970\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950931\t5984781\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950931\t5985922\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950931\t5987282\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950931\t5993506\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950931\t5994299\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950931\t5973056\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950931\t\n",
      "ENSG00000125885\t1\t5950652\t5998977\t5950931\t5967878\n",
      "ENSG00000129355\t-1\t10566460\t10569059\t10566460\t10568513\n",
      "ENSG00000129355\t-1\t10566460\t10569059\t10566460\t10567058\n",
      "ENSG00000129355\t-1\t10566460\t10569059\t10566601\t\n",
      "ENSG00000129355\t-1\t10566460\t10569059\t10566601\t10568513\n",
      "ENSG00000129355\t-1\t10566460\t10569059\t10566601\t10567058\n",
      "ENSG00000172201\t1\t19837370\t19842197\t19837370\t19837755\n",
      "ENSG00000172201\t1\t19837370\t19842197\t19837370\t19838584\n",
      "ENSG00000172201\t1\t19837370\t19842197\t19837370\t\n",
      "ENSG00000183434\t-1\t133216662\t133218354\t133216662\t133217042\n",
      "ENSG00000204086\t1\t96883908\t96885467\t96883908\t96884311\n",
      "ENSG00000206557\t1\t32817997\t32897824\t32817997\t32818081\n",
      "ENSG00000206557\t1\t32817997\t32897824\t32817997\t32873818\n",
      "ENSG00000206557\t1\t32817997\t32897824\t32817997\t32885934\n",
      "ENSG00000206557\t1\t32817997\t32897824\t32817997\t32890360\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#human_dataset.show_filters()\n",
    "#human_dataset.show_attributes()\n",
    "\n",
    "entrylist = [\"ENSG00000123080\", \"ENSG00000116574\", \"ENSG00000125885\", \"ENSG00000129355\", \n",
    "             \"ENSG00000183434\", \"ENSG00000172201\", \"ENSG00000206557\", \"ENSG00000204086\"]\n",
    "\n",
    "\"\"\"\n",
    "response = human_dataset.search({\n",
    "  'filters' : {'ensembl_gene_id' : entrylist, 'upstream_flank' : 100},\n",
    "  'attributes' : ['ensembl_gene_id', 'strand', 'start_position', 'end_position', 'transcript_start', \n",
    "                  'genomic_coding_start', 'gene_flank']\n",
    "  },\n",
    "  header=1)\n",
    "print(response.text)\n",
    "\"\"\"\n",
    "\n",
    "# Zonder upstream flank\n",
    "response = human_dataset.search({\n",
    "  'filters' : {'ensembl_gene_id' : entrylist},\n",
    "  'attributes' : ['ensembl_gene_id', 'strand', 'start_position', 'end_position', 'transcript_start', \n",
    "                  'genomic_coding_start']\n",
    "  },\n",
    "  header=1)\n",
    "print(response.text)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
